{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C, the allelic measure of deviation from Lande, for your chosen squared\n",
      " effect size distribution is: 0.6840765649154079\n",
      "\n",
      "\n",
      "The steady-state standard deviation in phenotype distribution for your chosen\n",
      "parameters in units little delta = root(Vs/(2N)) is: 12.73393504719654\n",
      "To be in the parameter regime considered in the paper, you want it to be >>1,\n",
      "and << root(2N)= 44.721359549995796\n",
      "\n",
      "\n",
      "The shift/root(Vs) is: 0.5694788877319691\n",
      "To be in the parameter regime considered in the paper, \n",
      "you want it to on the order of 1 or smaller\n",
      "\n",
      "\n",
      "The simulations (AA and OA) that run using an approximation for E[dx] (which\n",
      "assumes that integrals of certain functions over the phenotype distribution\n",
      "can be approximated by the function value at the mean phenotype) are not\n",
      "valid unless 1 >> VA/Vs, which at steady-state for your parameter\n",
      "choices is 0.08107655089311018\n",
      "ALL the trajectory simulations (OA) use this assumption\n",
      "as do the population simulations just follow alleles (AA).\n",
      "Thus if this condition is not satisfied, ONLY the population simulattions\n",
      "that simulate the FULL MODEL will be accurate (don't use the others) \n"
     ]
    }
   ],
   "source": [
    "# CELL A\n",
    "# YOU MUST ALWAYS RUN THIS CELL BEFORE\n",
    "# RUNNING ANY OTHERS\n",
    "# If you run all the cells in order, they will work.\n",
    "# Once you have produced and summarized some simulation\n",
    "# results using cells B-G\n",
    "# You could just run this cell and then skip to cell H\n",
    "# to play around wit plotting\n",
    "import os\n",
    "from combined_theory import get_C, get_variance_units_little_del_square\n",
    "\n",
    "# NBNBNB CHANGE my_base_directory_local\n",
    "# to be the directory in which you place the folder\n",
    "# called \"code\"\n",
    "my_base_directory_local = \"directory_contain_poly1D/PolygenicAdaptation1D\"\n",
    "\n",
    "my_base_directory_local = '/Users/Alexandra/pcloud_synced/PycharmProjects/Polished_code_python3/PolygenicAdaptation1D'\n",
    "\n",
    "save_output_dir =  os.path.join(my_base_directory_local,'results_local_test_output')\n",
    "save_sims_directory = os.path.join(my_base_directory_local,'results_local_test')\n",
    "save_sims_land_directory =  os.path.join(my_base_directory_local,'results_local_test_lande')\n",
    "code_directory = os.path.join(my_base_directory_local,\"code\")\n",
    "\n",
    "# the population size is 1000\n",
    "N = 1000\n",
    "# the shift size in units of initial phenotypic SD is 2\n",
    "D_s0 = 2\n",
    "# the number of mutations per gamete per gen\n",
    "U = 0.02\n",
    "# expected scaled selection coeff of new muts\n",
    "# (squared effects have a gamma distribution)\n",
    "E2Ns = 10\n",
    "# If you don't choose the variance in scaled\n",
    "#selection coefficient it defaults to E2Ns^2\n",
    "# making the distribution of incoming squared\n",
    "#effects exponential\n",
    "\n",
    "sigma_0_del = get_variance_units_little_del_square(E2Ns,mut_input=2*N*U)**(0.5)\n",
    "C = get_C(E2Ns)\n",
    "shift_over_root_Vs = D_s0*sigma_0_del/float(2*N)**(0.5)\n",
    "print(\"C, the allelic measure of deviation from Lande, for your chosen squared\\n\"\n",
    "      \" effect size distribution is: \"+str(C))\n",
    "print(\"\\n\")\n",
    "print(\"The steady-state standard deviation in phenotype distribution for your chosen\\n\"\n",
    "      \"parameters in units little delta = root(Vs/(2N)) is: \" + str(sigma_0_del))\n",
    "print(\"To be in the parameter regime considered in the paper, you want it to be >>1,\\n\"\n",
    "      \"and << root(2N)= \"+ str((2.0*N)**(0.5)))\n",
    "print(\"\\n\")\n",
    "print(\"The shift/root(Vs) is: \" + str(shift_over_root_Vs))\n",
    "print(\"To be in the parameter regime considered in the paper, \"\n",
    "      \"\\nyou want it to on the order of 1 or smaller\")\n",
    "print(\"\\n\")\n",
    "print(\"The simulations (AA and OA) that run using an approximation for E[dx] (which\\n\"\n",
    "      \"assumes that integrals of certain functions over the phenotype distribution\\n\"\n",
    "      \"can be approximated by the function value at the mean phenotype) are not\\n\"\n",
    "      \"valid unless 1 >> VA/Vs, which at steady-state for your parameter\\n\"\n",
    "      \"choices is \" + str(sigma_0_del**2/float(2*N)))\n",
    "print(\"ALL the trajectory simulations (OA) use this assumption\\n\"\n",
    "      \"as do the population simulations just follow alleles (AA).\")\n",
    "print(\"Thus if this condition is not satisfied, ONLY the population simulattions\\n\"\n",
    "      \"that simulate the FULL MODEL will be accurate (don't use the others) \")\n",
    "\n",
    "# save the params in a dict\n",
    "# so that later we can locate the folder\n",
    "# with the simulation results, and run OA\n",
    "# trajectory simulations using the D(t)\n",
    "# from those results\n",
    "param_dict = dict()\n",
    "param_dict['N'] = N\n",
    "param_dict['U'] = U\n",
    "param_dict['shift_s0'] = D_s0\n",
    "param_dict['E2Ns'] = E2Ns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "data": {
      "text/plain": "0"
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CELL B\n",
    "# This cell will take about 10-15 min to\n",
    "# run for the params chosen in Cell A.\n",
    "# Recommend running this >=5 times\n",
    "# to get reasonable plots\n",
    "# And many, many more times if you want the trajectory sims (OA)\n",
    "# that use the average D(t) from these sims to be accurate\n",
    "\n",
    "# This cell simulates the populations\n",
    "# The default parameters are chosen, so that the sims\n",
    "# will not take long to test on your computer\n",
    "# In particular you will want a better\n",
    "# realistic burn time, lag time, population size\n",
    "# and mutation rate\n",
    "# Maybe about 10 min for the params chosen in Cell A\n",
    "\n",
    "code_file = os.path.join(code_directory,\"simulate_populations_argparse.py\")\n",
    "\n",
    "# the population size is N\n",
    "arguments = \" -N \"+str(param_dict['N'])\n",
    "# the shift size in units of initial phenotypic SD is D_s0\n",
    "arguments += \" -D_s0 \" + str(param_dict['shift_s0'])\n",
    "# the number of mutations per gamete per gen is U\n",
    "arguments += \" -U \" + str(param_dict['U'])\n",
    "# expected scaled selection coeff of new muts is E2Ns\n",
    "arguments += \" -E2Ns \"+str(param_dict['E2Ns'])\n",
    "\n",
    "# the -hi flag means that histogram stats are saved (recommend)\n",
    "# the -r 2 means that the simulator will do two runs\n",
    "# of evolution\n",
    "# I set it to 2 rather then 1, so you will have\n",
    "# something to test the code that summarizes the results of\n",
    "# multiple runs of evolution\n",
    "# If there is only one run, there is nothing\n",
    "# to summarize.\n",
    "# In reality you will want to simulate many more than\n",
    "# 3 runs of evolution\n",
    "arguments += \" -hi -r 2 -sD \" + save_sims_directory\n",
    "# arguments += \" -a exact \"\n",
    "run_in_terminal = \"python \" + code_file + arguments\n",
    "\n",
    "# saves the standard error and standard output to text files,\n",
    "# which you can look at if something goes wrong\n",
    "run_in_terminal += ' > stdout.txt 2> stderror.txt'\n",
    "\n",
    "# Running the code via the terminal\n",
    "os.system(run_in_terminal)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "data": {
      "text/plain": "0"
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#CELL C\n",
    "# Run this before the next cell\n",
    "# Text results wil be saved in\n",
    "# results_local_test_output/text_output\n",
    "# inside your chosen my_base_directory_local\n",
    "\n",
    "# This cell summarizes the results and saves the summaries\n",
    "# using summarize_save_argparse.py\n",
    "# It also deletes the individual runs. If you have run time-consuming\n",
    "# simulations with the full population, as opposed to the All Alleles version,\n",
    "# you may not want to delete the individual runs\n",
    "# This also saves some results to a text file\n",
    "\n",
    "code_file = os.path.join(code_directory,\"summarize_save_argparse.py\")\n",
    "arguments = \" -rD \"+ save_sims_directory + \" -sD \" + save_output_dir\n",
    "run_in_terminal = \"python \" + code_file + arguments\n",
    "\n",
    "# saves the standard error and standard output to text files,\n",
    "# which you can look at if something goes wrong\n",
    "run_in_terminal += ' > stdout.txt 2> stderror.txt'\n",
    "\n",
    "# Running the code via the terminal\n",
    "os.system(run_in_terminal)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "You need to simulate a fairly large number of alleles to get reasonable results. In the paper we did 250 000.\n",
      "Note that when you simulate too few alleles, the 1.96*se (standard error) will NOT provide good confidence\n",
      "intervals as the central limit does not apply\n",
      "NOTE: YOU WILL NOT GET ACCURATE RESULTS WHEN LANDE = FALSE, UNLESS YOU HAVE RUN (and summarized) ENOUGH FULL POPULATION SIMS\n",
      " TO HAVE A NOT TOO LARGE STANDARD ERROR IN D(T)\n",
      "(or mu_3(t)/2V_A (t) which we use as a proxy for D(t) in the equilibration period)\n",
      "You need to simulate a fairly large number of alleles to get reasonable results. In the paper we did 250 000.\n",
      "Note that when you simulate too few alleles, the 1.96*se (standard error) will NOT provide good confidence\n",
      "intervals as the central limit does not apply\n"
     ]
    }
   ],
   "source": [
    "#CELL D\n",
    "#YOU must first have run Cell B Cell C\n",
    "# After you run this cell, you must run Cell G before plotting the results\n",
    "# from this cell\n",
    "\n",
    "# This cell will take ~10 min for the parameters\n",
    "# chosen in Cell A\n",
    "\n",
    "# This cell simulates allele trajectories using D(t) from\n",
    "# population simulations\n",
    "# The default parameters are silly, so that the sims\n",
    "# will not take too long to test on your computer\n",
    "\n",
    "# if Lande is true, the the sims will run with the Lande version of D(t),\n",
    "print(\"NOTE: YOU WILL NOT GET ACCURATE RESULTS WHEN LANDE = FALSE, UNLESS YOU HAVE RUN (and summarized)\"\n",
    "      \" ENOUGH FULL POPULATION SIMS\\n TO HAVE A NOT TOO LARGE STANDARD ERROR IN D(T)\\n(or mu_3(t)/2V_A (t) \"\n",
    "      \"which we use as a proxy for D(t) in the equilibration period)\")\n",
    "lande = False\n",
    "\n",
    "# Let's run the program a few times so there is something\n",
    "# to summarize\n",
    "num_runs = 4\n",
    "\n",
    "# save_sims_directory needs to contain sims corresponding to an\n",
    "# actual simulation folder with population simulation results\n",
    "# This is because (if lande=False) we are reading D(t) from\n",
    "# these simulations\n",
    "# If Lande is True, we are reading the phenotypic variance\n",
    "# and the shift size from these simulations, but using the\n",
    "# the corresponding Lande D(t)\n",
    "\n",
    "code_file = os.path.join(code_directory,\"simulate_trajectories_argparse.py\")\n",
    "\n",
    "arguments = \" -psF \" + save_sims_directory\n",
    "# the -nM 250 means that each tuple has 10 alleles\n",
    "# In reality you would want to simulate many more than 250 alleles for\n",
    "# each, say, effect size to get reasonable estimates of statistics\n",
    "# corresponding to allele trajectories of alleles with that effect.\n",
    "# I just chose parameters so the sims will run not too slowly for testing\n",
    "arguments += \" -nM 250 \"\n",
    "\n",
    "# if it exists, the program will find the simulation folder,\n",
    "# in psF=save_sims_directory for the population simulations\n",
    "# that you previously ran. It will read D(t)\n",
    "# from those sims and simulate trajectories using that\n",
    "# D(t). In the Lande case, it will calculate sigma_0_del\n",
    "# for those parameters and run simulations with the\n",
    "# corresponding Lande D(t)\n",
    "\n",
    "# the population size is N\n",
    "arguments += \" -N \"+str(param_dict['N'])\n",
    "# the shift size in units of initial phenotypic SD is D_s0\n",
    "arguments += \" -D_s0 \" + str(param_dict['shift_s0'])\n",
    "# the number of mutations per gamete per gen is U\n",
    "arguments += \" -U \" + str(param_dict['U'])\n",
    "# expected scaled selection coeff of new muts is E2Ns\n",
    "arguments += \" -E2Ns \"+str(param_dict['E2Ns'])\n",
    "if lande:\n",
    "    arguments+= ' -l'\n",
    "run_in_terminal = \"python \" + code_file + arguments\n",
    "\n",
    "# saves the standard error and standard output to text files,\n",
    "# which you can look at if something goes wrong\n",
    "run_in_terminal += ' > stdout.txt 2> stderror.txt'\n",
    "\n",
    "# Running the code num_runs times via the terminal\n",
    "for _ in range(num_runs):\n",
    "    os.system(run_in_terminal)\n",
    "\n",
    "print(\"You need to simulate a fairly large number of alleles to get reasonable results. In the paper we did 250 000.\\n\"\n",
    "  \"Note that when you simulate too few alleles, the 1.96*se (standard error) will NOT provide good confidence\\nintervals \"\n",
    "  \"as the central limit does not apply\")\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Your chosen steady-state standard deviation in phenotype distribution\n",
      "in units little delta = root(Vs/(2N)) is: 20\n",
      "To be in the parameter regime considered in the paper, you want it to be >>1,\n",
      "and << root(2N)= 100.0\n",
      "\n",
      "\n",
      "The shift/root(Vs) is: 0.4\n",
      "To be in the parameter regime considered in the paper, \n",
      "you want it to on the order of 1 or smaller\n",
      "\n",
      "\n",
      "The trajectory simulations (OA) run using an approximation for E[dx] (which\n",
      "assumes that integrals of certain functions over the phenotype distribution\n",
      "can be approximated by the function value at the mean phenotype) are not\n",
      "valid unless 1 >> VA/Vs, which at steady-state for your parameter\n",
      "choices is 0.04\n",
      "Thus if this condition is not satisfied, do not run trajectory simulations.\n",
      "As the results will not be valid \n"
     ]
    }
   ],
   "source": [
    "#Cell E\n",
    "# Setting the parameters for the PURE LANDE trajectoories (OA)\n",
    "# (in Cell F)\n",
    "# For these trajectories you don't need a preexisting simulation folder with simulation results, from which\n",
    "\n",
    "# the steady state phenotypic standard deviation is 20 in units little delta\n",
    "sigma_0_del = 20\n",
    "# the population size is 5000\n",
    "N = 5000\n",
    "# the shift size in units of initial phenotypic SD is 2\n",
    "D_s0 = 2\n",
    "\n",
    "shift_over_root_Vs = D_s0*sigma_0_del/float(2*N)**(0.5)\n",
    "print(\"Your chosen steady-state standard deviation in phenotype distribution\\n\"\n",
    "      \"in units little delta = root(Vs/(2N)) is: \" + str(sigma_0_del))\n",
    "print(\"To be in the parameter regime considered in the paper, you want it to be >>1,\\n\"\n",
    "      \"and << root(2N)= \"+ str((2.0*N)**(0.5)))\n",
    "print(\"\\n\")\n",
    "print(\"The shift/root(Vs) is: \" + str(shift_over_root_Vs))\n",
    "print(\"To be in the parameter regime considered in the paper, \"\n",
    "      \"\\nyou want it to on the order of 1 or smaller\")\n",
    "print(\"\\n\")\n",
    "print(\"The trajectory simulations (OA) run using an approximation for E[dx] (which\\n\"\n",
    "      \"assumes that integrals of certain functions over the phenotype distribution\\n\"\n",
    "      \"can be approximated by the function value at the mean phenotype) are not\\n\"\n",
    "      \"valid unless 1 >> VA/Vs, which at steady-state for your parameter\\n\"\n",
    "      \"choices is \" + str(sigma_0_del**2/float(2*N)))\n",
    "print(\"Thus if this condition is not satisfied, do not run trajectory simulations.\\n\"\n",
    "      \"As the results will not be valid \")"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [],
   "source": [
    "# Cell F\n",
    "# Should take ~20 min with parameters chosen in cell E\n",
    "# First run cell E which specifies your parameter choices\n",
    "# You must run Cell G before plotting the results\n",
    "# from this one\n",
    "\n",
    "# This cell simulates the PURE Lande trajectories. So unlike the previous run of trajectory\n",
    "# simulations you don't need a preexisting simulation folder with simulation results, from which\n",
    "# the simulator reads parameters (or possibly also D(t)).\n",
    "# Instead, you specify a population size, shift size and steady-state\n",
    "# phenotypic standard deviation (in units little delta(=sqrt(Vs/2N))).\n",
    "# The simulator then simulates trajectories of alleles with the\n",
    "# corresponding Lande D(t)\n",
    "# The default parameters (specifically the nM the number of mutations)\n",
    "# are chosen so that the sims will not take long to test on your computer\n",
    "\n",
    "# Let's run the program a few times so there is something\n",
    "# to summarize\n",
    "num_runs = 5\n",
    "\n",
    "code_file = os.path.join(code_directory,\"simulate_trajectories_argparse.py\")\n",
    "\n",
    "# the steady state phenotypic standard deviation is sigma_0_del in units little delta\n",
    "arguments = \" -sigma_0_del \"+str(sigma_0_del)\n",
    "# the population size is N\n",
    "arguments += \" -N \"+str(N)\n",
    "# the shift size in units of initial phenotypic SD is D_s0\n",
    "arguments += \" -D_s0 \"+str(D_s0)\n",
    "# as beforethe -nM 100 means that each tuple has 100 mutations\n",
    "# of evolution\n",
    "\n",
    "# ONLY SIMULATING with 100 mutatins per effect\n",
    "# size. You definitely want to use more than this\n",
    "arguments +=  ' -nM 100'\n",
    "# It is '-oL' that indicates we are running PURE Lande\n",
    "arguments += \" -oL\"\n",
    "# specify the folder name to save pure Lande traj results\n",
    "arguments += \" -lF \" + save_sims_land_directory\n",
    "run_in_terminal = \"python \" + code_file + arguments\n",
    "\n",
    "# saves the standard error and standard output to text files,\n",
    "# which you can look at if something goes wrong\n",
    "run_in_terminal += ' > stdout.txt 2> stderror.txt'\n",
    "\n",
    "# Running the code num_runs times via the terminal\n",
    "for _ in range(num_runs):\n",
    "    os.system(run_in_terminal)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [],
   "source": [
    "# Cell G\n",
    "# You must run this cell after you have run\n",
    "# either Cell D or Cell E&F, to summarize\n",
    "# those results\n",
    "# Text results wil be saved in\n",
    "# results_local_test_output/text_output\n",
    "# inside your chosen my_base_directory_local\n",
    "\n",
    "# This cell summarizes the results and saves the summaries\n",
    "# Also deletes the individual runs\n",
    "# Also saves some results to a text file\n",
    "\n",
    "# Lande is true if you are summarising pure Lande trajectory\n",
    "# results\n",
    "# Otherwise lande=False\n",
    "lande_list =[True,False]\n",
    "\n",
    "code_file = os.path.join(code_directory,\"summarize_save_argparse.py\")\n",
    "arguments = \" -rD \"+ save_sims_directory + \" -sD \" + save_output_dir\n",
    "arguments_Lande = \" -rD \"+ save_sims_land_directory + \" -sD \" + save_output_dir\n",
    "run_in_terminal = \"python \" + code_file + arguments\n",
    "run_in_terminal_Lande = \"python \" + code_file + arguments_Lande\n",
    "\n",
    "# Uncomment below if you need to see the output and/or standard error\n",
    "run_in_terminal += ' > redirected.txt 2> stderror.txt'\n",
    "for lande in lande_list:\n",
    "    if not lande:\n",
    "        os.system(run_in_terminal)\n",
    "    else:\n",
    "        os.system(run_in_terminal_Lande)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'N': 1000, 'Vs': 2000.0, 'shift_s0': 2.0, 'var_0': 162.15310178621888, 'sigma_0_del': 12.733935047196482, 'shift': 25.467870094392964, 'U': 0.02, 'E2Ns': 10.0, 'V2Ns': 100.0, 'scale_s': 10.0, 'shape_s': 1.0, 'lag_time': 1000, 'T_max': 10901, 'algorithm': 'approx', 't_freeze_new_muts_start': 0, 't_length_freeze_new_muts': 5000, 't_freeze': 0, 'burn_time': 5000}\n",
      "{'N': 1000, 'Vs': 2000.0, 'shift_s0': 2.0, 'var_0': 162.15310178621888, 'sigma_0_del': 12.733935047196482, 'shift': 25.467870094392964, 'U': 0.02, 'E2Ns': 10.0, 'V2Ns': 100.0, 'scale_s': 10.0, 'shape_s': 1.0, 'lag_time': 1000, 'T_max': 10901, 'algorithm': 'approx', 't_freeze_new_muts_start': 0, 't_length_freeze_new_muts': 5000, 't_freeze': 0, 'burn_time': 5000, 'var_0_emp': 173.31332001788564, 'var_0_emp_se': 4.73427527075523}\n",
      "final_stats  not in traj stats\n",
      "reading param dict\n",
      "{'N': 5000, 'Vs': 10000.0, 'shift_s0': 2.0, 'var_0': 400.0, 'sigma_0_del': 20.0, 'shift': 40.0, 'time_from_which_use_dist_guess': 0, 'var_0_emp': 400.0}\n",
      "final_stats  not in traj stats\n"
     ]
    }
   ],
   "source": [
    "#Cell H\n",
    "###PLOTTING SETUP#######\n",
    "# Always run this cell before running any of the\n",
    "# cells below to plot figures.\n",
    "# Figures will be saved in\n",
    "# results_local_test_output/figures\n",
    "# inside your chosen my_base_directory_local.\n",
    "import os\n",
    "from summarize_save_class import SumAndSave\n",
    "\n",
    "# This cell just sets us up to plot a few very rough figures\n",
    "# whic we do in the following two cells\n",
    "\n",
    "save_figure_directory = os.path.join(save_output_dir, \"figures\")\n",
    "\n",
    "# parent directory is the directory were you results directories are stored\n",
    "# save_output_dir is the directory you want new output to be saved\n",
    "#You must have run population sims, with cell B\n",
    "# and summarized them with Cell C first\n",
    "my_sum_and_save= SumAndSave(parent_dir=save_sims_directory,save_output_dir=save_output_dir)\n",
    "\n",
    "#You must have run pure lande trajectory sims, with Cell E&F\n",
    "# and summarized them with Cell G first\n",
    "my_sum_and_save_Land = SumAndSave(parent_dir=save_sims_land_directory,save_output_dir=save_output_dir)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "#Cell I\n",
    "# First run Cell H to set up sum_and_saves for plotting\n",
    "#You also must have run population sims, with cell B\n",
    "# and summarized them with Cell C\n",
    "\n",
    "# The code that does the plotting is messy and hard to understand\n",
    "# You can also use the text files to generate whatever plots you wish.\n",
    "# This cell allows you to quickly look at some results from the\n",
    "# full population simulations\n",
    "\n",
    "# This cell plots some basic statistics like the\n",
    "# distance of the mean phenotype from the optimum (dist)\n",
    "# or the skewness or variance of the phenotype distribution\n",
    "# dist_guess = third moment/(2*variance) of\n",
    "# the phenotype distribution\n",
    "\n",
    "# The saved plots appear in results_local_test_output/figures/approx/parameter_folder\n",
    "# 'approx' indicates that the approximate population sims (AA) were used\n",
    "# to generate the sim results, and 'parameter_folder' has the names of\n",
    "# the various paraters used in the text\n",
    "\n",
    "from plot_class import PlotClass\n",
    "\n",
    "\n",
    "bstat_list = ['skewness','opt_minus_mean_fixed_over_shift','dist_guess']\n",
    "bstat_list += ['d2ax_frozen_over_shift','dist_sum_over_2N_scaled']\n",
    "\n",
    "#better to look at these stats with trait in units\n",
    "# of initial phenotypic SD\n",
    "bstat_list_unts_s0 = ['var','dist']\n",
    "\n",
    "end_time = 5000\n",
    "first_time = 0\n",
    "\n",
    "for data_class_param_dict in my_sum_and_save.param_dict_list():\n",
    "    myparamdictkey = my_sum_and_save.turn_dict_into_dict_key(data_class_param_dict)\n",
    "    datafolder = my_sum_and_save.make_folder_name_for_data_class(myparamdictkey)\n",
    "    current_save_figure_directory = os.path.join(save_figure_directory,datafolder)\n",
    "    tmp_dclass = my_sum_and_save.get_data_class(myparamdictkey)\n",
    "    parameter_string = tmp_dclass.make_parameter_text_to_write_file()\n",
    "    current_save_figure_directory = os.path.join(current_save_figure_directory,parameter_string)\n",
    "    if not os.path.exists(current_save_figure_directory):\n",
    "        os.makedirs(current_save_figure_directory)\n",
    "\n",
    "    # if units are true we plot\n",
    "    # in units of steady-state SD\n",
    "    # if units are false we plot\n",
    "    # in units of little delta (=sqrt(Vs/2N))\n",
    "    tmp_dclass.set_units(False)\n",
    "\n",
    "    tmp_plots = PlotClass([tmp_dclass],base_dir=current_save_figure_directory)\n",
    "\n",
    "    for bstati in bstat_list:\n",
    "        tmp_plots.plot_bstat(bstat=bstati,std=False,domain=[first_time ,end_time])\n",
    "\n",
    "    tmp_dclass.set_units(True)\n",
    "    tmp_plots = PlotClass([tmp_dclass],base_dir=current_save_figure_directory)\n",
    "    for bstati in bstat_list_unts_s0:\n",
    "        tmp_plots.plot_bstat(bstat=bstati,std=False,domain=[first_time ,end_time])\n",
    "        tmp_dclass.set_units(True)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C 0.6840765649154079\n"
     ]
    }
   ],
   "source": [
    "# Cell J\n",
    "# First run Cell H to set up sum_and_saves for plotting\n",
    "# You also must have run population sims, with cell B\n",
    "# and summarized them with Cell C\n",
    "# It will only produce figures if you ran the population sims\n",
    "# with the -hi flag\n",
    "# Figures saved in 'hbstats_at_times' folder\n",
    "# Results likely not very good unless you have run many pop sims\n",
    "\n",
    "# This cell plots some binned stats for the (AA or full model)\n",
    "# population sims (binned by effect size, or initial MAF)\n",
    "from plot_class import PlotClass\n",
    "\n",
    "\n",
    "poses = ['both','pos','neg']\n",
    "#Most interesting to look at aligned and opposing combined\n",
    "poses = ['both']\n",
    "bins = ['_efs_bins','_fbins']\n",
    "\n",
    "statsplotall = ['frozen_d2ax_scaled_per_mut_input',\n",
    "                'frozen_d2ax_over_shift',\n",
    "                'frozen_nm_2ax_over_shift',\n",
    "                'frozen_nm_2ax_scaled_per_mut_input',\n",
    "                'frozen_numseg_per_mut_input',\n",
    "                'frozen_nm_and_standing_d2ax_scaled_per_mut_input']\n",
    "\n",
    "units = False\n",
    "\n",
    "maxt = 10000\n",
    "\n",
    "for data_class_param_dict in my_sum_and_save.param_dict_list():\n",
    "    myparamdictkey = my_sum_and_save.turn_dict_into_dict_key(data_class_param_dict)\n",
    "    datafolder = my_sum_and_save.make_folder_name_for_data_class(myparamdictkey)\n",
    "    current_save_figure_directory = os.path.join(save_figure_directory,datafolder)\n",
    "    tmp_dclass = my_sum_and_save.get_data_class(myparamdictkey)\n",
    "\n",
    "    lande_time = my_sum_and_save.get_lande_time(myparamdictkey)\n",
    "\n",
    "    # we can only make these plots if we saved histogram stats for the sims\n",
    "    if tmp_dclass._hbstats:\n",
    "        parameter_string = tmp_dclass.make_parameter_text_to_write_file()\n",
    "        current_save_figure_directory = os.path.join(current_save_figure_directory,parameter_string)\n",
    "        if not os.path.exists(current_save_figure_directory):\n",
    "            os.makedirs(current_save_figure_directory)\n",
    "        tmp_plots = PlotClass([tmp_dclass],base_dir=current_save_figure_directory)\n",
    "        tmp_dclass.set_units(units)\n",
    "        for posi in poses:\n",
    "            for stati in statsplotall:\n",
    "                for bin in bins:\n",
    "                    nami = stati+bin\n",
    "                    if nami in tmp_dclass._hbstats:\n",
    "                        times = [-1,lande_time]\n",
    "                        tmp_plots.plot_h_bstat_times_hist(index = 0,h_bstat=nami,times = times,pos=posi)\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['frozen_nm_and_standing_d2ax_scaled_per_mut_input_efs_bins', 'frozen_d2ax_scaled_per_mut_input_fbins', 'frozen_d2ax_scaled_per_mut_input_efs_bins', 'frozen_nm_2ax_over_shift_efs_bins', 'frozen_d2ax_over_shift_fbins', 'frozen_numseg_per_mut_input_efs_bins', 'frozen_nm_2ax_scaled_per_mut_input_efs_bins', 'frozen_d2ax_over_shift_efs_bins', 'frozen_numseg_per_mut_input_fbins']\n"
     ]
    }
   ],
   "source": [
    "# Cell K\n",
    "# First run Cell H to set up sum_and_saves for plotting\n",
    "#You also must have run population sims, with cell B\n",
    "# and summarized them with Cell C\n",
    "# It will only produce figures if you ran the population sims\n",
    "# with the -hi flag\n",
    "# Figures saved in 'hbstats' folder\n",
    "# Results likely not very good unless you have run many pop sims\n",
    "\n",
    "# This cell plots some binned stats for the full population sims\n",
    "# (binned by effect size, or initial minor allele frequency)\n",
    "from plot_class import PlotClass\n",
    "\n",
    "poses = ['both','pos','neg']\n",
    "bins = ['_efs_bins','_fbins']\n",
    "\n",
    "statsplotall = ['frozen_d2ax_scaled_per_mut_input',\n",
    "                'frozen_d2ax_over_shift','frozen_nm_2ax_scaled_per_mut_input','frozen_numseg_per_mut_input']\n",
    "\n",
    "units = False\n",
    "\n",
    "maxt = 10000\n",
    "\n",
    "for data_class_param_dict in my_sum_and_save.param_dict_list():\n",
    "    myparamdictkey = my_sum_and_save.turn_dict_into_dict_key(data_class_param_dict)\n",
    "    datafolder = my_sum_and_save.make_folder_name_for_data_class(myparamdictkey)\n",
    "    current_save_figure_directory = os.path.join(save_figure_directory,datafolder)\n",
    "    tmp_dclass = my_sum_and_save.get_data_class(myparamdictkey)\n",
    "\n",
    "    # we can only make these plots if we saved histogram stats for the sims\n",
    "    if tmp_dclass._hbstats:\n",
    "        parameter_string = tmp_dclass.make_parameter_text_to_write_file()\n",
    "        current_save_figure_directory = os.path.join(current_save_figure_directory,parameter_string)\n",
    "        if not os.path.exists(current_save_figure_directory):\n",
    "            os.makedirs(current_save_figure_directory)\n",
    "        tmp_dclass.set_units(units)\n",
    "        tmp_plots = PlotClass([tmp_dclass],base_dir=current_save_figure_directory)\n",
    "\n",
    "        print(tmp_dclass._hbstats)\n",
    "\n",
    "        for posi in poses:\n",
    "            for stati in statsplotall:\n",
    "                for bin in bins:\n",
    "                    nami = stati+bin\n",
    "                    if nami in tmp_dclass._hbstats:\n",
    "                        tmp_plots.plot_h_bstat_over_time_quick(index = 0,h_bstat=nami,domain=[0,maxt],pos=posi)\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'N': 5000, 'Vs': 10000.0, 'shift_s0': 2.0, 'var_0': 400.0, 'sigma_0_del': 20.0, 'shift': 40.0}\n",
      "['2a_frac_fixed_scaled_per_mut_input', 'NUM_SEGREGATING', 'frac_fixed_per_seg_var', '2a_frac_fixed_per_mut_input', 'NUM_MUTANTS']\n",
      "we are in Lande\n",
      "['d2ax_scaled_per_mut_input']\n",
      "yes data\n",
      "after _read_traj_bstats\n",
      "we are in Lande\n",
      "['x_per_seg_var']\n",
      "yes data\n",
      "after _read_traj_bstats\n",
      "we are in Lande\n",
      "['x_per_seg_var']\n",
      "yes data\n",
      "after _read_traj_bstats\n",
      "yval is zero for S  100.0\n",
      "yval is zero for S  56.23413251903491\n",
      "yval is zero for S  100.0\n",
      "yval is zero for S  56.23413251903491\n",
      "yval is zero for S  100.0\n",
      "yval is zero for S  31.622776601683793\n",
      "yval is zero for S  56.23413251903491\n",
      "{'N': 1000, 'Vs': 2000.0, 'shift_s0': 2.0, 'var_0': 162.15310178621888, 'sigma_0_del': 12.733935047196482, 'shift': 25.467870094392964}\n",
      "['2a_frac_fixed_scaled_per_mut_input', 'NUM_SEGREGATING', 'frac_fixed_per_seg_var', '2a_frac_fixed_per_mut_input', 'NUM_MUTANTS']\n",
      "no lande\n",
      "['d2ax_scaled_per_mut_input']\n",
      "yes data\n",
      "no lande\n",
      "['x_per_seg_var']\n",
      "yes data\n",
      "no lande\n",
      "['x_per_seg_var']\n",
      "yes data\n",
      "yval is zero for S  100.0\n",
      "yval is zero for S  100.0\n",
      "yval is zero for S  100.0\n",
      "yval is zero for S  31.622776601683793\n",
      "yval is zero for S  56.23413251903491\n"
     ]
    }
   ],
   "source": [
    "#Cell L\n",
    "# This run plots results of trajectory simulations (all the\n",
    "# previous ones did population sims)\n",
    "# Plots for the end of the rapid phase\n",
    "# saved in 'tstats_times'\n",
    "# And long-term plots saved in 'ftstats'\n",
    "# First run Cell H to set up sum_and_saves for plotting\n",
    "\n",
    "sum_and_save_list = [my_sum_and_save]\n",
    "# To run this cell with 'my_sum_and_save' in the\n",
    "# sum_and_save_list you need to have run population sims, with cell B\n",
    "# and summarized them with Cell C\n",
    "#  You also need to have run trajectory sims\n",
    "# cell D and summarized with cell G\n",
    "\n",
    "sum_and_save_list = [my_sum_and_save_Land]\n",
    "# To run this cell with 'my_sum_and_save_Land' in the\n",
    "# sum_and_save_list you need to have run pur Lande trajectory sims\n",
    "# with Cell E&F and summarized them with Cell G\n",
    "from plot_class import PlotClass\n",
    "\n",
    "\n",
    "sum_and_save_list = [my_sum_and_save_Land,my_sum_and_save]\n",
    "#sum_and_save_list = [my_sum_and_save_Land]\n",
    "#sum_and_save_list = [my_sum_and_save]\n",
    "\n",
    "tstat_list_both = ['d2ax_scaled_per_mut_input']\n",
    "tstat_list_pos_neg = ['x_per_seg_var']\n",
    "\n",
    "final_stat_list_both = ['2a_frac_fixed_scaled_per_mut_input']\n",
    "final_stat_list_pos_neg = ['frac_fixed_per_seg_var']\n",
    "\n",
    "for a_sum_save_class in sum_and_save_list:\n",
    "    for data_class_param_dict in a_sum_save_class.param_dict_list():\n",
    "        myparamdictkey = a_sum_save_class.turn_dict_into_dict_key(data_class_param_dict)\n",
    "        tmp_dclass = a_sum_save_class.get_data_class(myparamdictkey)\n",
    "        datafolder = a_sum_save_class.make_folder_name_for_data_class(myparamdictkey)\n",
    "        current_save_figure_directory = os.path.join(save_figure_directory,datafolder)\n",
    "        parameter_string = tmp_dclass.make_parameter_text_to_write_file()\n",
    "        current_save_figure_directory = os.path.join(current_save_figure_directory,parameter_string)\n",
    "        tmp_plots = PlotClass([tmp_dclass],base_dir=current_save_figure_directory)\n",
    "        print(data_class_param_dict)\n",
    "        print(tmp_dclass._ftstats)\n",
    "        lande_time = a_sum_save_class.get_lande_time(myparamdictkey)\n",
    "        #lande_time = 1\n",
    "\n",
    "        tmp_dclass.set_units(False)\n",
    "        for stat in tstat_list_both:\n",
    "            tmp_plots.paper_plot_tstats_times(tstat=stat,indexes=[0],time=lande_time,compare=True)\n",
    "        for stat in tstat_list_pos_neg:\n",
    "            for pos in ['pos','neg']:\n",
    "                tmp_plots.paper_plot_tstats_times(tstat=stat,indexes=[0],time=lande_time,pos=pos)\n",
    "\n",
    "        tmp_dclass.set_units(False)\n",
    "        for stat in final_stat_list_both:\n",
    "            tmp_plots.paper_plot_final_tstats(ftstat=stat,indexes=[0],compare=True)\n",
    "        for stat in final_stat_list_pos_neg:\n",
    "            for pos in ['pos','neg']:\n",
    "                tmp_plots.paper_plot_final_tstats(ftstat=stat,indexes=[0],pos=pos)\n",
    "\n",
    "\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}